### PR TITLE
fix: boolean and/or expressions with null

### LINKED ISSUE #3512: `daft.Expression.__or__` and `daft.Expression.__and__` should not propagate nulls
### Describe the bug

Our expressions for or + and  yield null when either side is null, but when the statement is always true or false regardless of the null-side value, we should yield that instead of propagating nulls.

### To Reproduce

Consider this dataframe:
```
import daft

df = daft.from_pydict({
    "a": [True, False],
    "b": [None, None]
})
```

`df.select(df["a"] | df["b"])` and `df.select(df["a"] & df["b"])` both yield a column with two null rows.

Same behavior exists in SQL

### Expected behavior

`df.select(df["a"] | df["b"])` should yield [True, null] since the first row is always true.

`df.select(df["a"] & df["b"])` should yield [null, False] since the first row is always false.

### Component(s)

SQL, Other

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kevinzwang**
Resolves #3512

A fun truth table ($V_{x}$ is the validity $x$, $L$ is left, and $R$ is right):

| $L$ | $V_{L}$ | $R$ | $V_{R}$ | $∧$ | $V_{∧}$ | $∨$ | $V_{∨}$ |
| :-----: | :-----: | :-----: | :-----: | :-----: | :-----: | :-----: | :-----: |
| ⬜️ | ⬜️ | ⬜️ | ⬜️ |     | ⬜️ |     | ⬜️ |
| ⬜️ | ⬜️ | ⬜️ | ✅  | ⬜️ | ✅  |     | ⬜️ |
| ⬜️ | ⬜️ | ✅  | ⬜️ |     | ⬜️ |     | ⬜️ |
| ⬜️ | ⬜️ | ✅  | ✅  |     | ⬜️ | ✅  | ✅  |
| ⬜️ | ✅  | ⬜️ | ⬜️ | ⬜️ | ✅  |     | ⬜️ |
| ⬜️ | ✅  | ⬜️ | ✅  | ⬜️ | ✅  | ⬜️ | ✅  |
| ⬜️ | ✅  | ✅  | ⬜️ | ⬜️ | ✅  |     | ⬜️ |
| ⬜️ | ✅  | ✅  | ✅  | ⬜️ | ✅  | ✅  | ✅  |
| ✅  | ⬜️ | ⬜️ | ⬜️ |     | ⬜️ |     | ⬜️ |
| ✅  | ⬜️ | ⬜️ | ✅  | ⬜️ | ✅  |     | ⬜️ |
| ✅  | ⬜️ | ✅  | ⬜️ |     | ⬜️ |     | ⬜️ |
| ✅  | ⬜️ | ✅  | ✅  |     | ⬜️ | ✅  | ✅  |
| ✅  | ✅  | ⬜️ | ⬜️ |     | ⬜️ | ✅  | ✅  |
| ✅  | ✅  | ⬜️ | ✅  | ⬜️ | ✅  | ✅  | ✅  |
| ✅  | ✅  | ✅  | ⬜️ |     | ⬜️ | ✅  | ✅  |
| ✅  | ✅  | ✅  | ✅  | ✅  | ✅  | ✅  | ✅  |

### CODE REVIEW

File: tests/series/test_comparisons.py
Comment (kevinzwang) [inline tests/series/test_comparisons.py:1]: I only touched this file to update test values but Ruff wanted to reformat it as well

**Review by andrewgazelka — CHANGES_REQUESTED** [review]

File: src/daft-core/src/array/ops/comparison.rs [hunk 1]
```diff
@@ -601,13 +601,45 @@ impl Not for &BooleanArray {
...
+                arrow2::array::BooleanArray::new(
+                    arrow2::datatypes::DataType::Boolean,
+                    values.clone(),
+                    Some(new_validity),
+                ),
+            ))
+        }
+
         match (self.len(), rhs.len()) {
             (x, y) if x == y => {
```
[the remark this answers is not in the document]
Comment (kevinzwang) [reply src/daft-core/src/array/ops/comparison.rs:624]: This naming seems to exist across all of the ops so I think I will just leave it be. Don't want to make this PR too big

File: src/daft-core/src/array/ops/comparison.rs [hunk 2]
```diff
@@ -640,13 +672,45 @@ impl DaftLogical<&Self> for BooleanArray {
...
```
[the remark this answers is not in the document]
Comment (kevinzwang) [reply src/daft-core/src/array/ops/comparison.rs:695]: see above

**Comment by kevinzwang** [comment]
@andrewgazelka could you give this a second look?

**Review by andrewgazelka — APPROVED** [review]

**Pull request merged**